operator/controllers/resources/role_controller_test.go

```diff
@@ -33,57 +39,189 @@ import (
 	resourcesv5 "github.com/gravitational/teleport/operator/apis/resources/v5"
 )
 
+// When I create a TeleportRole CR in Kubernetes, the corresponding TeleportRole must be created in Teleport
+// When I delete a TeleportRole CR in Kubernetes, the corresponding TeleportRole must be deleted in Teleport
 func TestRoleCreation(t *testing.T) {
-	ctx := context.Background()
-
-	teleportServer, operatorName := defaultTeleportServiceConfig(t)
-
-	require.NoError(t, teleportServer.Start())
-
-	tClient := clientForTeleport(t, teleportServer, operatorName)
-	k8sClient := startKubernetesOperator(t, tClient)
-
-	ns := createNamespaceForTest(t, k8sClient)
+	ctx, tClient, k8sClient, ns := setupKubernetesAndTeleport(t)
 	roleName := validRandomResourceName("role-")
 
-	// The role is created in K8S
+	// End of setup, we create the role in Kubernetes
 	k8sCreateDummyRole(ctx, t, k8sClient, ns.Name, roleName)
 
+	// We wait for the role to be created in Teleport
 	fastEventually(t, func() bool {
 		tRole, err := tClient.GetRole(ctx, roleName)
 		if trace.IsNotFound(err) {
 			return false
 		}
 		require.NoError(t, err)
 
+		// Role should have the same name, and have the Kubernetes origin label
 		require.Equal(t, tRole.GetName(), roleName)
-
 		require.Contains(t, tRole.GetMetadata().Labels, types.OriginLabel)
 		require.Equal(t, tRole.GetMetadata().Labels[types.OriginLabel], types.OriginKubernetes)
 
 		return true
 	})
 
-	// The role is deleted in K8S
+	// Cleanup and test, we delete the role in Kubernetes```
 	k8sDeleteRole(ctx, t, k8sClient, roleName, ns.Name)
 
+	// We wait for the role to be deleted in Teleport
 	fastEventually(t, func() bool {
 		_, err := tClient.GetRole(ctx, roleName)
 		return trace.IsNotFound(err)
 	})
 }
 
-func TestRoleDeletionDrift(t *testing.T) {
-	ctx := context.Background()
+func TestRoleCreationFromYAML(t *testing.T) {
+	ctx, tClient, k8sClient, ns := setupKubernetesAndTeleport(t)
+	tests := []struct {
+		name         string
+		roleSpecYAML []byte
+		shouldFail   bool
+		expectedSpec *types.RoleSpecV5
+	}{
+		{
+			"Valid login list",
+			[]byte(`
+allow:
+  logins:
+  - ubuntu
+  - root
+`),
+			false,
+			&types.RoleSpecV5{
+				Allow: types.RoleConditions{
+					Logins: []string{"ubuntu", "root"},
+				},
+			},
+		},
```

Comment:
nit: Use labeled struct initialization? It has better readability and is easier to extend later.

```suggestion
		{
			name: "Valid login list",
			roleSpecYAML: []byte(`
allow:
  logins:
  - ubuntu
  - root
`),
			shouldFail: false,
			expectedSpec: &types.RoleSpecV5{
				Allow: types.RoleConditions{
					Logins: []string{"ubuntu", "root"},
				},
			},
		},
```

Same for others.